### PR TITLE
[RemoteMirror] Fix demangle node corruption caused by excessively eager clearing of the NodeFactory.

### PR DESCRIPTION
**Pull request by mikeash**
We clear the NodeFactory to prevent unbounded buildup of allocated memory, but this is done too eagerly. In particular, normalizeReflectionName can end up clearing the factory while the calling code is still using nodes that were allocated from it.

To keep peak memory usage low while avoiding this problem, we introduce a checkpoint mechanism in NodeFactory. A checkpoint can be pushed and then subsequently popped. When a checkpoint is popped, only the nodes allocated since the checkpoint was pushed are invalidated and the memory reclaimed. This allows us to quickly clear short-lived nodes like those created in normalizeReflectionName, while preserving longer-lived nodes used in code calling it. Uses of clearNodeFactory are replaced with this checkpoint mechanism.

rdar://106547092

### CODE REVIEW

**Comment by mikeash** [comment]
@swift-ci please test macos platform

File: stdlib/public/RemoteInspection/TypeRefBuilder.cpp [hunk 1]
```diff
@@ -93,6 +93,7 @@ RemoteRef<char> TypeRefBuilder::readTypeRef(uint64_t remoteAddr) {
 /// Load and normalize a mangled name so it can be matched with string equality.
 llvm::Optional<std::string>
 TypeRefBuilder::normalizeReflectionName(RemoteRef<char> reflectionName) {
+  auto checkpoint = pushNodeFactoryCheckpoint();
```
Comment (al45tair) [inline stdlib/public/RemoteInspection/TypeRefBuilder.cpp:96]: A lot of code that does this seems to follow the pattern
```
auto checkpoint = pushNodeFactoryCheckpoint();
...
popNodeFactoryCheckpoint(checkpoint);
```
which makes me wonder whether we should have a
```
template <typename T, typename F>
T withScopedNodes(F func) {
  auto checkpoint = pushNodeFactoryCheckpoint();
  T result = func();
  popNodeFactoryCheckpoint(checkpoint);
  return result;
}
```
or similar (obviously that's written OTOH in a Github comment, but you get the idea).

**Review by al45tair — APPROVED** [review]
LGTM.

File: stdlib/public/RemoteInspection/TypeRefBuilder.cpp (hunk 1 above)
Comment (mikeash) [reply stdlib/public/RemoteInspection/TypeRefBuilder.cpp]: RAII would be more idiomatic in C++. I've gone ahead and added a little RAII wrapper and switched over to that. The scoping makes things less convenient in a couple of places where we declare locals that get used later, but overall it's nicer.

Comment (al45tair) [reply stdlib/public/RemoteInspection/TypeRefBuilder.cpp]: :-) I seems to have Swift on the brain at the moment. Yes, RAII is more idiomatic in C++.

**Comment by mikeash** [comment]
@swift-ci please test

**Pull request merged**